Feature/Sync-GitRepository

jeffprice6 opened the pull request.

splatteredbits commented on LibGit2/Functions/Sync-GitRepository.ps1:
```diff
@@ -0,0 +1,74 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+function Sync-GitRepository
```
Rename to `Send-GitCommit`. `Sync-GitRepository` would be used if the function also handles pulling down any new changes as part of pushing.

jeffprice6 replied:
Updated. This naming convention makes sense.

splatteredbits commented on Tests/Sync-GitRepository.Tests.ps1:
```diff
@@ -0,0 +1,128 @@
+    GivenCommittedChangeToPush
+    WhenSynchronizingGitRepository
+    ThenErrorWasThrown 'No upstream remote is configured for ''master'' branch. Aborting synchronization.'
+}
```
What happens if there are changes in the remote that haven't been pulled down? Add a test that ensures the correct behavior. Does `$repo.Network.Push` return an object? If not, should Send-GitCommit return `$true` if it pushes and `$false` otherwise?

jeffprice6 replied:
Updated functionality and added test cases.

jeffprice6 commented:
I've just pushed some updates based upon earlier feedback. This one should be good for review.

splatteredbits commented on LibGit2/Functions/Send-GitCommit.ps1:
```diff
@@ -0,0 +1,89 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+function Send-GitCommit
```
This function should return an enumeration value based on what happened:

````
public enum PushResult
{
    Ok,
    Failed,
    Rejected
}
```

The way this is currently written, any automation that tries to push will fail when there are incoming changes. Callers need to know if the push is something they can handle. In the case where the push is rejected because there are new remote changes, automation can use that information to pull those changes down, then re-attempt the push.

jeffprice6 replied:
Added output enumerator.

splatteredbits commented on Tests/Send-GitCommit.Tests.ps1:
```diff
@@ -0,0 +1,197 @@
+    GivenRemoteContainsOtherChanges
+    GivenCommittedChangeToPush
+    WhenSendingGitCommits
+    ThenErrorWasThrown 'Cannot push because a reference that you are trying to update on the remote contains commits that are not present locally.'
```
This should be

```
ThenNoErrors
ThenResultIs [LibGit2.Automation.PushResult]::Rejected
```

splatteredbits commented on Tests/Send-GitCommit.Tests.ps1:
```diff
@@ -0,0 +1,197 @@
+    GivenLocalRepositoryTracksRemote 'LocalRepo'
+    WhenSendingGitCommits
+    ThenNoErrorsWereThrown
+    ThenWarningWasWritten 'There are no outgoing commits to push to the remote repository.'
```
No warning. Instead,

```
ThenResultIs [LibGit2.Automation.PushResult]::Ok
```

splatteredbits commented on LibGit2/Functions/Send-GitCommit.ps1:
```diff
@@ -0,0 +1,89 @@
+        }
+        else
+        {
+            Write-Warning -Message 'There are no outgoing commits to push to the remote repository.'
```
No warning.

jeffprice6 commented:
I've updated the function to include assignment of output enum status based on the type of error and updated the tests. This one should be ready for review.. :)
